File: hash/rust/rosetta-image-hash/src/hex.rs
//! Hex string ↔ Hash conversion (square shape and 14×B colorhash shape).

use crate::internal::bitpack;
use crate::{Hash, ImageHashError};

pub fn hex_to_hash(hex: &str) -> Result<Hash, ImageHashError> {
    let bits = bitpack::unpack_square(hex)?;
    Ok(Hash::from_bits_unchecked(bits))
}

pub fn hex_to_flathash(hex: &str, hash_size: usize) -> Result<Hash, ImageHashError> {
    if hash_size < 1 {
        return Err(ImageHashError::InvalidBinbits(hash_size));
    }
    let bits = bitpack::unpack_flat(hex, hash_size)?;
    Ok(Hash::from_bits_unchecked(bits))
}

File: hash/js/rosetta-squint-hash/src/hex.ts
import { Hash, ImageHashError } from "./hash.js";
import { unpackFlat, unpackSquare } from "./internal/bitpack.js";

export function hexToHash(hex: string): Hash {
  const bits = unpackSquare(hex);
  return new Hash(bits);
}

export function hexToFlathash(hex: string, hashSize: number): Hash {
  if (hashSize < 1) {
    throw new ImageHashError("InvalidBinbits", `hashSize must be >= 1, got ${hashSize}`);
  }
  const bits = unpackFlat(hex, hashSize);
  return new Hash(bits);
}
